### PR TITLE
Document the Error types in std::string

### PR DESCRIPTION
**Pull request by steveklabnik**
Part of #29376

### CODE REVIEW

File: src/libcollections/string.rs [hunk 1]
```diff
@@ -36,7 +36,41 @@ pub struct String {
     vec: Vec<u8>,
 }
 
-/// A possible error value from the `String::from_utf8` function.
+/// A possible error value when converting a `String` from a UTF8 byte vector.
```
Comment (apasel422) [inline src/libcollections/string.rs]: s/UTF8/UTF-8/

File: src/libcollections/string.rs [hunk 2]
```diff
@@ -44,7 +78,26 @@ pub struct FromUtf8Error {
     error: Utf8Error,
 }
 
-/// A possible error value from the `String::from_utf16` function.
+/// A possible error value when converting a `String` from a UTF16 byte slice.
```
Comment (apasel422) [inline src/libcollections/string.rs]: s/UTF16/UTF-16/

File: src/libcollections/string.rs [hunk 3]
```diff
@@ -337,8 +390,8 @@ impl String {
     ///
     /// ```
     /// // 𝄞music
-    /// let mut v = &mut [0xD834, 0xDD1E, 0x006d, 0x0075,
-    ///                   0x0073, 0x0069, 0x0063];
+    /// let v = &mut [0xD834, 0xDD1E, 0x006d, 0x0075,
```
Comment (alexcrichton) [inline src/libcollections/string.rs]: I think in these examples the `mut` here can be dropped as well (may have just bee included by accident long ago)

File: src/libcollections/string.rs [hunk 4]
```diff
@@ -44,7 +78,26 @@ pub struct FromUtf8Error {
     error: Utf8Error,
 }
 
-/// A possible error value from the `String::from_utf16` function.
+/// A possible error value when converting a `String` from a UTF-16 byte slice.
+///
+/// This type is the error type for the [`from_utf16()`] method on [`String`].
+/// It provides no specific details on the kind of error, just that one
+/// occurred.
```
Comment (alexcrichton) [inline src/libcollections/string.rs]: This seems a little odd to document like this, the intention here is to have an opaque type that we can then modify in the future to have more internals if necessary, so can this sentence perhaps just be dropped?

Comment (steveklabnik) [inline src/libcollections/string.rs]: Sure thing

File: src/libcollections/string.rs (hunk 3 above)
Comment (steveklabnik) [inline src/libcollections/string.rs]: :+1:

**Pull request merged**